### PR TITLE
Asynchronous DNS for sockets

### PR DESCRIPTION
**Pull request by gvilums**
### What does this PR do?
Changes usockets to resolve DNS queries asynchronously through a global resolver, instead of blocking.

### How did you verify your code works?
Existing tests.

### CODE REVIEW

File: packages/bun-usockets/src/loop.c [hunk 1]
```diff
@@ -165,25 +169,42 @@ void us_internal_handle_low_priority_sockets(struct us_loop_t *loop) {
     }
 }
 
+void us_internal_handle_dns_results(struct us_loop_t *loop) {
+    pthread_mutex_lock(&loop->data.mutex);
+    struct us_connecting_socket_t *s = loop->data.dns_ready_head;
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/loop.c]: the mutex shouldn't be necessary, we should be able to do an atomic swap

File: packages/bun-usockets/src/internal/networking/bsd.h [hunk 2]
```diff
@@ -40,6 +40,7 @@
 #endif
 /* For socklen_t */
 #include <sys/socket.h>
+#include <netdb.h>
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/internal/networking/bsd.h:43]: was this include intentional

File: packages/bun-usockets/src/internal/loop_data.h [hunk 3]
```diff
@@ -18,6 +18,9 @@
 #ifndef LOOP_DATA_H
 #define LOOP_DATA_H
 
+#include <pthread.h>
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/internal/loop_data.h]: if we use atomics we can avoid the mutex and this include (which is probably causing the windows build failure)

File: packages/bun-usockets/src/internal/internal.h [hunk 4]
```diff
@@ -122,6 +126,20 @@ struct us_socket_t {
...
+    alignas(LIBUS_EXT_ALIGNMENT) void *addrinfo_req;
+    struct us_socket_context_t *context;
+    struct us_connecting_socket_t *next;
+    struct us_socket_t *socket;
+    int options;
+    int socket_ext_size;
+    unsigned int closed : 1, shutdown : 1, ssl : 1, shutdown_read : 1;
+    unsigned char timeout;
+    unsigned char long_timeout;
+    int error;
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/internal/internal.h:155]: does the error need to be stored?

File: packages/bun-usockets/src/context.c [hunk 5]
```diff
@@ -343,33 +345,88 @@ struct us_listen_socket_t *us_socket_context_listen_unix(int ssl, struct us_sock
...
 #ifndef LIBUS_NO_SSL
     if (ssl) {
-        return (struct us_socket_t *) us_internal_ssl_socket_context_connect((struct us_internal_ssl_socket_context_t *) context, host, port, source_host, options, socket_ext_size);
+        return us_internal_ssl_socket_context_connect((struct us_internal_ssl_socket_context_t *) context, host, port, options, socket_ext_size);
     }
 #endif
 
-    LIBUS_SOCKET_DESCRIPTOR connect_socket_fd = bsd_create_connect_socket(host, port, source_host, options);
+    struct us_connecting_socket_t *c = malloc(sizeof(struct us_connecting_socket_t) + socket_ext_size);
+    memset(c, 0, sizeof(struct us_connecting_socket_t));
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/context.c]: instead of malloc and then memset 0, we can use calloc

File: packages/bun-usockets/src/context.c [hunk 6]
```diff
@@ -343,33 +345,88 @@ struct us_listen_socket_t *us_socket_context_listen_unix(int ssl, struct us_sock
...
+struct us_connecting_socket_t *us_socket_context_connect(int ssl, struct us_socket_context_t *context, const char *host, int port, int options, int socket_ext_size) {
 #ifndef LIBUS_NO_SSL
     if (ssl) {
-        return (struct us_socket_t *) us_internal_ssl_socket_context_connect((struct us_internal_ssl_socket_context_t *) context, host, port, source_host, options, socket_ext_size);
+        return us_internal_ssl_socket_context_connect((struct us_internal_ssl_socket_context_t *) context, host, port, options, socket_ext_size);
     }
 #endif
 
-    LIBUS_SOCKET_DESCRIPTOR connect_socket_fd = bsd_create_connect_socket(host, port, source_host, options);
+    struct us_connecting_socket_t *c = malloc(sizeof(struct us_connecting_socket_t) + socket_ext_size);
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/context.c]:
```suggestion
    struct us_connecting_socket_t *c = us_calloc(1, sizeof(struct us_connecting_socket_t) + socket_ext_size);
```

File: packages/bun-usockets/src/context.c [hunk 7]
```diff
@@ -19,6 +19,8 @@
 #include "internal/internal.h"
 #include <stdlib.h>
 #include <string.h>
+#include <pthread.h>
+#include <stdio.h>
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/context.c]: did you mean to include stdio.h?

File: packages/bun-usockets/src/bsd.c [hunk 8]
```diff
@@ -905,7 +905,7 @@ static int bsd_do_connect(struct addrinfo *rp, int *fd)
             return LIBUS_SOCKET_ERROR;
         }
 
-        int resultFd = bsd_create_socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+        int resultFd = bsd_create_socket(rp->ai_family, SOCK_STREAM, 0);
```
Comment (Jarred-Sumner) [inline packages/bun-usockets/src/bsd.c]: why hardcode it?

Comment (gvilums) [reply packages/bun-usockets/src/bsd.c]: these are the only valid values for a TCP socket, and makes `bsd_do_connect` independent of the values returned by `getaddrinfo` beyond the resolved addresses

File: packages/bun-usockets/src/internal/internal.h (hunk 4 above)
Comment (gvilums) [reply packages/bun-usockets/src/internal/internal.h:155]: it's necessary so `us_connecting_socket_get_error` can return the right error if `getaddrinfo` fails, as `getaddrinfo` doesn't set `errno`

File: src/http.zig [hunk 9]
```diff
@@ -742,13 +742,14 @@ pub const HTTPThread = struct {
...
-        const loop = bun.uws.Loop.create(struct {
-            pub fn wakeup(_: *uws.Loop) callconv(.C) void {
-                http_thread.drainEvents();
-            }
-            pub fn pre(_: *uws.Loop) callconv(.C) void {}
-            pub fn post(_: *uws.Loop) callconv(.C) void {}
-        });
+        // const loop = bun.uws.Loop.create(struct {
+        //     pub fn wakeup(_: *uws.Loop) callconv(.C) void {
+        //         http_thread.drainEvents();
```
Comment (Jarred-Sumner) [inline src/http.zig]: Where is `drainEvents` called?

**Comment by Jarred-Sumner** [comment]
Looks like WebSocket (websocket_http_client) needs to be updated

File: src/http.zig (hunk 9 above)
Comment (gvilums) [reply src/http.zig]: Before every iteration of the event loop (`processEvents`)

**Pull request merged**